Fix particle initial adaptive execution

gassmoeller opened the pull request:
This prevents particles from being advected when the 'Run postprocessors on initial refinement = true' option is set. Before the particles would have been unintentionally advected with the flow.
Apart from that the particle postprocessor now behaves like the visualization postprocessor, only writing files once at the first initial adaptive refinement step to not mess up the naming scheme of the output files. Beware that this means for an otherwise unchanged model that setting 'Run postprocessors on initial refinement = true' will change the particle output in the first timestep, because it outputs the state *before* the adaptive refinement instead of *after* the initial refinement (as it will also do for the visualization output).

bangerth commented:
Need to run ./doc/indent.

bangerth commented on include/aspect/simulator_access.h:
```diff
@@ -256,6 +256,17 @@ namespace aspect
+       * This can be useful for plugins that want to function differently in
+       * the initial adaptive refinements and later on.
+       * This will be not initialized before Simulator<dim>::run() is called.
+       * It equals 0 to parameters.initial_adaptive_refinement during the
```
"It equals" is a bit hard to understand. How about "It iterates upward from 0 to ..."?

bangerth commented on source/postprocess/tracers.cc:
```diff
@@ -93,8 +93,12 @@ namespace aspect
+          && (this->get_pre_refinement_step() < this->get_parameters().initial_adaptive_refinement);
+      if (!in_initial_refinement)
+        // Advance the particles in the world to the current time
+        world.advance_timestep();
```
Yes, this looks right -- only do it after the last refinement step. Do I read this right?

bangerth replied:
I'm confused because I think I understand what this code does, but that doesn't seem to match what you write in your introduction to the patch:
........................
...only writing files once at the first initial adaptive refinement step to not mess up the naming scheme of the output files. Beware that this means for an otherwise unchanged model that setting 'Run postprocessors on initial refinement = true' will change the particle output in the first timestep, because it outputs the state before the adaptive refinement instead of after the initial refinement (as it will also do for the visualization output).
......
The code only seems to run the advancement after the solution has been computed after the *last* of the pre-refinement steps. When you say "only writing files once at the first initial adaptive refinement step" I guess you mean "once" as in "only one time" as opposed to "after a condition has been met", and the "first initial adaptive refinement step" is in fact the "first time step, which may consist of multiple sub-steps separated by refinement"?

Also, why does the output change?

gassmoeller replied:
Maybe I formulated that in a misleading way: The *advection* of the particles now only happens after the last refinement step so everything is correct with that (so this PR fulfils its purpose). 
During coding I noticed this peculiarity: The *output* of the first timestep is only written after the first refinement step (the first call to execute() will call the `output->output_particle_data(...)` function). After the next refinement when we call execute() the `last_output_time` will have been updated and therefore we will return early from this function (line 107). I think this behaviour is a bit strange, but it is the same in the visualization plugin, so I left it in its current state.
I do not know exactly how to explain the output change best, maybe a sketch helps. Lets assume you have 3 initial adaptive refinements, then with 'Run postprocessors on initial refinement = false' the output will work like this:
1. refinement.
2. refinement.
3. refinement <---- write output files (only call of the postprocessors)

With 'Run postprocessors on initial refinement = true' the output will work like this:

1. refinement <---- write output files (first call of postprocessrs)
2. refinement <---- output is skipped (early return from function)
3. refinement <---- output is skipped (early return from function)

If we want to change this we should change it in both the visualization and tracers postprocessor to stay consistent, but I am not sure if we should, and it is not related to this PR.

bangerth commented:
This is in response to a comment in #791.

bangerth commented:
Ah, I see. In your flow chart,
```
Lets assume you have 3 initial adaptive refinements, then with 'Run postprocessors on initial refinement = false' the output will work like this:
1. refinement.
2. refinement.
3. refinement <---- write output files (only call of the postprocessors)

With 'Run postprocessors on initial refinement = true' the output will work like this:

refinement <---- write output files (first call of postprocessrs)
refinement <---- output is skipped (early return from function)
refinement <---- output is skipped (early return from function)
```
I suppose that after every refinement and before writing output (or skipping output), there is a "solution" phase?

I agree that the behavior is awkward. At the same time, the option is probably rarely used. We could delay generating output in the same way as we delay advecting particles now. Feel free to open a github issue for this.

tjhei commented:
This might be a bug that got introduced a long time ago when we reworked how we determine when to write output (I remember we changed from storing "next output time" to "last output time" or something). "Run postprocessors on initial refinement" should run tings after every adaptive step.

gassmoeller commented:
I think I have addressed all comments so this is ready for another review or merge. I have opened an issue for the other thing (#802). I will try to come back to that one later.
